### DISCUSSION TITLE
Next js 13 Server/Client Error for ClassName

### DISCUSSION THREAD

**Discussion by aydinvivik** [pmndrs/zustand#1802] [Q&A]
Hello, I am using zustand for the first time. I get an error when defining a class according to the value I get from the state.

**state.tsx;**

```
import { create } from "zustand";

type AppActions = {
  setSidebar: (payload: any) => void
}

const initialStates = {
  sidebar: typeof localStorage !== "undefined" ? !!localStorage.getItem('sidebar-folded') : false,
  sidebarMenu: false,
}

export type AppState = typeof initialStates;

const useAppStore = create<AppState & AppActions>((set) => ({
  ...initialStates,
  setSidebar: (payload: any) => set((state) => ({ sidebar: payload }))
}));

export default useAppStore;
```

**sidebar-component;**

```
import useAppStore from "@/store/useAppStore";

export default function Sidebar() {
  const sidebar = useAppStore(state => state.sidebar);
  const sidebarMenu = useAppStore(state => state.sidebarMenu);
  const isMobile = useMediaQuery('(max-width: 1200px');

  return (
    <aside className={`${styles.sidebar_wrapper} ${sidebar && !isMobile ? 'app-sidebar folded' : 'app-sidebar'}`}>
      <div className="app-sidebar-row app-sidebar-header">
        <Brand />
      </div>
      <div className="app-sidebar-row app-sidebar-body">
      <ScrollArea className="scroll-bar" scrollbarSize={6}>
        <SidebarMenu />
      </ScrollArea>
      </div>
      <div className="app-sidebar-row app-sidebar-footer">
        
      </div>
    </aside>
  )
}
```

**Warning: Prop `className` did not match. Server: "sidebar_sidebar_wrapper__0TzWU app-sidebar" Client: "sidebar_sidebar_wrapper__0TzWU app-sidebar folded"**

![Screen Shot 2023-05-12 at 04 20 55](https://github.com/pmndrs/zustand/assets/5020002/5c0adb75-56b3-45ff-ace1-a4b53e86ec06)

I tried with-zustand in the next js examples but I still had the same problem.

**Comment by dbritto-dev** [comment]
@aydinvivik this `const isMobile = useMediaQuery('(max-width: 1200px');` shuldn't be `const isMobile = useMediaQuery('(max-width: 1200px)');` ? 

BTW, for `isMobile` you should use `useState` so react can re-hydrate it correctly

**Reply by aydinvivik** [reply]
I think the problem is not in `isMobile`, when I use the `sidebar` without `isMobile` I still get the same error, but if I use only isMobile I don't get the error.

**Reply by aydinvivik** [reply]
There is no problem if I don't give an item in state as a class, I even like zustang a lot, but I don't know how to solve this problem.

**Reply by dbritto-dev** [reply]
Let me look into that

**Reply by aydinvivik** [reply]
Hi there, Have you found any solution?

**Reply by aydinvivik** [reply]
I think the problem is with localstorage, If I give a normal boolen value instead of the default value localstorage in the state, it does not give an error.


**before:**
```
const initialStates = {
  sidebar: typeof localStorage !== "undefined" ? !!localStorage.getItem('sidebar-folded') : false,
  sidebarMenu: false,
}
```

**after;**
```
const initialState: appState = {
  sidebar: false,
  sidebarMenu: false
}
```

but how can I make it get the state value from localstorage?

**Reply by Fawwaz-2009** [reply]
I'm seeing similar bug when I use `persist`

```
warning: Prop `disabled` did not match. Server: "" Client: "false" at button
 ```

**Reply by dbritto-dev** [reply]
@aydinvivik yeah, that's the expected behavior. On server value is `false` but then on client is `true` and those values doesn't match. 

What you can do is something like this:

```ts
const initialState: AppState = { 
  sidebar: false,
  sidebarMenu: false
}
```
then update `sidebar` on mount

```ts
const setSidebar = useAppStore(state => state.setSidebar)

useEffect(() => {
  setSidebar(!!localStorage.getItem('sidebar-folded'))
}, [])
```

> **Note:** for `hydration` to work, your initial render output must be identical on the client and the server.

**Reply by aydinvivik** [reply]
thank you for your answer, I did it this way because I thought it would be with useEffect, but this time when I refresh the page, when the page is reloaded, the sidebar appears normal and goes to the folded position with a delay of 1-2 seconds. I hope I won't encounter such a problem when I build the project. Despite everything, I like zustand a lot and I will continue to use it.

**Reply by dbritto-dev** [reply]
> thank you for your answer, I did it this way because I thought it would be with useEffect, but this time when I refresh the page, when the page is reloaded, the sidebar appears normal and goes to the folded position with a delay of 1-2 seconds. I hope I won't encounter such a problem when I build the project. Despite everything, I like zustand a lot and I will continue to use it.

Would you mind trying again with `useLayoutEffect`?

**Reply by aydinvivik** [reply]
I will try. Thank you.
